staging/src/k8s.io/dynamic-resource-allocation/api/types.go

```diff
@@ -0,0 +1,64 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package api
+
+import (
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type ResourceSlice struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+	Spec ResourceSliceSpec
+}
+
+type ResourceSliceSpec struct {
+	Driver       UniqueString
+	Pool         ResourcePool
+	NodeName     string
+	NodeSelector *v1.NodeSelector
+	AllNodes     bool
+	Devices      []Device
+}
+
+type ResourcePool struct {
+	Name               UniqueString
+	Generation         int64
+	ResourceSliceCount int64
+}
+type Device struct {
+	Name  UniqueString
+	Basic *BasicDevice
```

Comment:
This conversion from the external API into a representation that is used internally is also going to be useful for https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/4815-dra-partitionable-devices: that KEP introduces a new device type. It's possible to convert this `BasicDevice` into the new type here.  Then the scheduling code only needs to deal with that one.